src/elements/fields/FileUploadField.jsx

```diff
@@ -1,35 +1,218 @@
-import React from 'react';
+import React, { useRef } from 'react';
+import { useFileData, useThumbnailData } from '../../utils/image';
+import { justRemove } from '../../utils/array';
 
-import ReactForm from 'react-bootstrap/Form';
+import { FiX } from 'react-icons/fi';
+import { IconContext } from 'react-icons';
+import { Image } from 'react-bootstrap';
+import { FileUploadIcon } from '../components/icons';
 
-function FileUploadField({
+function RichFileUploadField({
   element,
-  fieldLabel,
   applyStyles,
   required = false,
   editable = false,
-  onChange = () => {},
-  onClick = () => {},
+  onChange: customOnChange = () => {},
+  onClick: customOnClick = () => {},
+  initialFiles = [],
   elementProps = {},
   children
 }) {
   const servar = element.servar;
+  const showLabel = servar.name !== '';
+  const isMultiple = servar.metadata.multiple;
+  const fileInput = useRef();
+
+  const [rawFiles, setRawFiles] = useFileData(initialFiles);
+  const thumbnailData = useThumbnailData(rawFiles);
+  // verify that thumbnail data still has length when there's no thumbnail for a text file
+  const allowMoreFiles = isMultiple || thumbnailData.length === 0;
+  const fileExists = thumbnailData.length > 0 || rawFiles.length > 0;
+
+  function onClick(event) {
+    if (!allowMoreFiles) return;
+    fileInput.current.click();
+    customOnClick(event);
+  }
+
+  // When the user uploads files to the multi-file upload, we just append to the existing set
+  // By default the input element would just replace all the uploaded files (we don't want that)
+  async function onChange(event) {
+    const uploadedFiles = Array.from(event.target.files).map((file) =>
+      Promise.resolve(file)
+    );
+    const newRawFiles = [...rawFiles, ...uploadedFiles];
+    setRawFiles(newRawFiles);
+    customOnChange(newRawFiles, rawFiles.length);
+
+    // Wipe the value of the upload element so we can upload multiple copies of the same file
+    // If we didn't do this, then uploading the same file wouldn't re-trigger onChange
+    fileInput.current.value = [];
+  }
+
+  function onClear(index) {
+    return () => {
+      const newRawFiles = justRemove(rawFiles, index);
+      setRawFiles(newRawFiles);
+      customOnChange(newRawFiles, index);
+    };
+  }
+
+  const imgStyles = {
+    maxWidth: '100%',
+    maxHeight: '100%',
+    minHeight: 0,
+    ...applyStyles.getTarget('img')
+  };
+  const usingDefaultIcon = !element.properties.icon;
+  // Default img width is '', which causes the default icon to have no width
+  // and not appear. Set a default width if the user hasn't set one
+  if (usingDefaultIcon && Number.isNaN(Number.parseInt(imgStyles.width))) {
```

Comment:
instead of setting a width just for the default icon, is there a css prop that basically says "expand to fill container width or height, whichever one is hit first". e.g. max-content, auto, etc.

Reply:
yeah I'll try that